[RFC 140 2b] `pkgs/by-name`: Automated migration

infinisil opened the pull request:
### Context

This is the [second part](https://github.com/NixOS/rfcs/blob/master/rfcs/0140-simple-package-paths.md#pr-2-automated-migration) of the implementation of the accepted [RFC 140](https://github.com/NixOS/rfcs/pull/140), the automated migration. The first part is already done and can be found here: https://github.com/NixOS/nixpkgs/pull/237439.

The plan for the second part can be found here: https://github.com/NixOS/nixpkgs/issues/258650

This is hopefully only the first of many quality-of-life improvements to Nixpkgs by the [Nixpkgs Architecture Team](https://nixos.org/community/teams/nixpkgs-architecture).

This PR is currently an old draft of the implementation, using [a slightly hacky tool](https://github.com/nixpkgs-architecture/nix-spp) to do it. Will be updated after part 1 is done.

## TODO

- [ ] Update the manual migration guidelines added in https://github.com/NixOS/nixpkgs/pull/254632

infinisil committed:
Implement unit directory reading

infinisil committed:
Fix evaluation from previous automatic migration

Tested with nix-env -qaf .

The res -> super changes are necessary because res points to the result
of the current overlay, but some of the packages were moved to the
previous overlay, which can be accessed with super. This is a
fundamental problem to be fixed, see https://github.com/NixOS/nixpkgs/pull/55061
and related issues

The schleuder change is necessary because the ./default.nix file
referenced ./., but there's also another package in ./bin. The tool
couldn't move the gemfiles in ./. because

The schleuder change is necessary because the tool isn't smart enough
and this is a special case of the default.nix reading gemfiles from a
./. reference and there being both schleuder in
pkgs/tools/security/schleuder and schleuder-cli in
pkgs/tools/security/schleuder/cli

zowoq commented:
`[skip editorconfig]` should help with this.

infinisil commented:
@zowoq Not entirely how that works, see [here](https://docs.github.com/en/actions/managing-workflow-runs/skipping-workflow-runs), but that can't disable just a specific workflow, it disables all of them which I don't want

zowoq commented:
https://github.com/NixOS/nixpkgs/blob/31ce2fc58405e1ef3cfc96755c4970716c5ece5a/.github/workflows/editorconfig.yml#L14

`[skip editorconfig]` skips only the editorconfig check

SuperSandro2000 commented on pkgs/top-level/stage.nix:
```diff
@@ -8,7 +8,15 @@
+  # reused every time this file is called
+  unitDirs = builtins.foldl' (result: shard:
+    result // builtins.mapAttrs (name: type:
+      ../unit + "/${shard}/${name}/pkg-fun.nix"
```
```suggestion
      ../unit + "/${shard}/${name}"
```
Why not keep default.nix as a filename?

infinisil replied:
See https://github.com/nixpkgs-architecture/simple-package-paths#alternate-pkg-funnix-filename, feel free to open issues in that repository for feedback :)

SuperSandro2000 replied:
https://github.com/nixpkgs-architecture/simple-package-paths/issues/30

infinisil committed:
documentation-highlighter: Only depend on needed files

Makes the build independency of the default.nix and update.sh file by
explicitly specifying the files that are needed in the result

This allows changing those files without causing a rebuild

infinisil committed:
zsh-clipboard: Make independent of Nix file

This way, changing the Nix file won't change the derivation

infinisil committed:
Run automatic migration

Using https://github.com/nixpkgs-architecture/nix-spp/commit/94a4f56ed13c1354c57369ebf7c2f8c014c30f6a

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2022-01-23-nixpkgs-architecture-team-meeting-26/24892/1

AndersonTorres commented on pkgs/top-level/stage.nix:
```diff
@@ -8,7 +8,15 @@
+  # reused every time this file is called
+  unitDirs = builtins.foldl' (result: shard:
+    result // builtins.mapAttrs (name: type:
+      ../unit + "/${shard}/${name}/pkg-fun.nix"
```
This implies a hardcoded limit of 2 in the subdirs?

infinisil replied:
Yes indeed, two levels means that `pkgs/unit` has a bit less than 1000 subdirectories, and each of those also has less than 1000 subdirectories with the current package distribution.

infinisil committed:
Update backport.yml

infinisil committed:
Testing backports for moved files

infinisil committed:
Merge pull request #3 from nixpkgs-architecture/pr-test

Testing backports for moved files

infinisil committed:
PR test 2

infinisil committed:
Merge pull request #4 from nixpkgs-architecture/pr-test-2

PR test 2

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2023-09-05-nixpkgs-architecture-team-meeting-43/32635/1

infinisil commented:
I'm planning to implement this in two steps:
1. Implement a CI check to require new packages to use `pkgs/by-name` if possible. This implements the exclusion list from [part 2](https://github.com/NixOS/rfcs/blob/master/rfcs/0140-simple-package-paths.md#pr-2-automated-migration), but doesn't migrate any packages yet.

   This could benefit from https://github.com/NixOS/nixpkgs/issues/256788
3. The mass migration itself, which now only needs to do three things for each top-level package using `callPackage`:
   - Move all of its files to `pkgs/by-name`
   - Remove it from the exclusion list
   - If it uses an empty second argument, remove its definition from `all-packages.nix`. This is also planned to be enforced with https://github.com/NixOS/nixpkgs/pull/256792

iFreilicht commented on .github/workflows/backport.yml:
```diff
@@ -17,7 +17,7 @@ jobs:
       contents: write # for korthout/backport-action to create branch
       pull-requests: write # for korthout/backport-action to create PR to backport
     name: Backport Pull Request
-    if: github.repository_owner == 'NixOS' && github.event.pull_request.merged == true && (github.event_name != 'labeled' || startsWith('backport', github.event.label.name))
```
I think this should be changed back after this PR has been merged, otherwise this workflow might run in every contributors repo as well, right?

infinisil replied:
Oh yeah I only used this for testing. I'll have to completely update this PR once everything is ready.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/ci-will-soon-enforce-pkgs-by-name-for-new-packages/38098/1

gaykitty commented:
FYI the [Nixpkgs Architecture Team](https://nixos.org/community/teams/nixpkgs-architecture.html) link is incorrect or dead.

infinisil commented:
Note that @Aleksanaa did a large part of the automated migration in https://github.com/NixOS/nixpkgs/pull/354531, thank you! There's still more that can be migrated manually, so I'll leave this PR open for now until I or somebody else can do it
